crates/sui-core/src/node_sync/node_state.rs

```diff
@@ -562,21 +563,31 @@ where
             })
             .tap_err(|e| error!(?digest, "error: {}", e))?;
 
-        self.process_parents(permit, &digests.transaction, &effects)
-            .await?;
+        self.process_parents(
+            Some(permit),
+            &digests.transaction,
+            &effects,
+            PARENT_WAIT_TIMEOUT,
+        )
+        .await?;
 
         self.state
-            .handle_node_sync_certificate(cert, effects.clone())
+            .handle_node_sync_certificate(&cert, &effects)
             .await?;
 
         Ok(SyncStatus::CertExecuted)
     }
 
-    async fn process_parents(
+    /// Transaction initialized by Node Sync passes in a Semaphore for
+    /// concurrency controlling while Transaction initialized by the write path
+    /// (TransactionOrchestrator) does not.
+    pub async fn process_parents<S>(
```

Comment:
Let's not make this method public if at all possible.

Reply:
You can already get the result of an execution via the existing interface. Here's how execution driver does it: https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority_active/execution_driver/mod.rs#L126